### PR TITLE
FS-1581: Assessor Task List page state endpoint

### PR DESCRIPTION
**Pull request by tferns**
**Summary:** 

- Add endpoint to retrieve state for populating the front-end assessor task list page
- Consumed by: https://github.com/communitiesuk/funding-service-design-assessment/pull/58

**Note:** Supersedes https://github.com/communitiesuk/funding-service-design-assessment-store/pull/45

### CODE REVIEW

File: tasks/db_tasks.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def seed_dev_db(c):
                 f" {apps * rounds * funds} test rows."
             )
 
-            seed_database(apps, rounds, funds)
+            seed_database_randomly(apps, rounds, funds)
```
Comment (gidsg) [inline tasks/db_tasks.py:101]: Why are we making this change in this PR? Seems unrelated?

File: db/schemas/schemas.py [hunk 2]
```diff
@@ -11,7 +12,47 @@ class AssessmentRecordMetadata(SQLAlchemyAutoSchema):
...
+        )
+
+    @post_dump
+    def discard_irrelevant_fields(self, item, many, **kwargs):
+        """Discard irrelevant fields from the jsonb_blob."""
+        item["date_submitted"] = item["jsonb_blob"]["date_submitted"]
+        item["project_reference"] = item["short_id"]
+        del item["short_id"]
+        del item["jsonb_blob"]
+        return item
```
Comment (SentientHousePlant) [inline db/schemas/schemas.py:58]: Why are we doing this?

File: db/schemas/schemas.py [hunk 3]
```diff
@@ -11,7 +12,47 @@ class AssessmentRecordMetadata(SQLAlchemyAutoSchema):
...
+    - round_id
+    """
+
+    class Meta:
+        model = AssessmentRecord
+        exclude = tuple(
+            field
+            for field in AssessmentRecord.__table__.columns.keys()
+            if field not in ASSESSOR_TASK_LIST_METADATA_FIELDS
+        )
```
Comment (SentientHousePlant) [inline db/schemas/schemas.py:49]: You can use `include` when you create the serialiser object - IE `serialiser = AssessmentRecordSchema(include=("short_id", "data_submitted"))`

Comment (SentientHousePlant) [reply db/schemas/schemas.py:49]: This way you don't have to write new objects for exclusion

File: db/schemas/schemas.py (hunk 2 above)
Comment (SentientHousePlant) [reply db/schemas/schemas.py:58]: NB: see key and field here https://marshmallow.readthedocs.io/en/stable/marshmallow.fields.html#marshmallow.fields.Field . This might be a cleaner way of doing this. Also, we extract information from the json blob on insert so we don't have to do it here

File: db/schemas/schemas.py (hunk 3 above)
Comment (SentientHousePlant) [reply db/schemas/schemas.py:49]: If we are missing a column add it and update the mapper here https://github.com/communitiesuk/funding-service-design-assessment-store/blob/main/db/queries/assessment_records/_helpers.py#L7

File: db/schemas/__init__.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
 from .schemas import AssessmentRecordMetadata
+from .schemas import AssessorTaskListMetadata
 
-__all__ = [AssessmentRecordMetadata]
+__all__ = ["AssessmentRecordMetadata", "AssessorTaskListMetadata"]
```
Comment (SentientHousePlant) [inline db/schemas/__init__.py:4]: 💯

File: api/routes/assessment_routes.py [hunk 5]
```diff
@@ -22,3 +25,28 @@ def all_assessments_for_fund_round_id(
...
+    :return: A dictionary.
+    """
+
+    metadata = find_assessor_task_list_state(application_id)
+    sections, criterias = transform_to_assessor_task_list_metadata(
+        metadata["fund_id"], metadata["round_id"]
+    )
+
+    # We don't need to return this state.  It's only used to get the config above.
+    del metadata["round_id"]
```
Comment (SentientHousePlant) [inline api/routes/assessment_routes.py:44]: `del` is not typically used in python for filter since it's not very explicit and can create issues because dictionarys are typically passed by reference rather then value - i'd recommand writing a dictionary comprehension which filters and creates a new filtered dict to return.

**Review by SentientHousePlant — CHANGES_REQUESTED** [review]

File: tasks/db_tasks.py (hunk 1 above)
Comment (tferns) [reply tasks/db_tasks.py:101]: Think it was re-named but not committed in another PR, I added it so I could run the script and committed it at the same time for anyone else running it (incase they hit the same issue)

File: db/schemas/schemas.py (hunk 3 above)
Comment (tferns) [reply db/schemas/schemas.py:49]: I was using that originally but then at some point decided to go with this, I think it's when I (at least at the time) thought I had to write a new class and couldn't achieve what I was trying purely by passing properties (the `@post_dump` thing) - I'll take a look at re-using `include` if I can remove the `@post_dump` utilizing the docs you mentioned above

File: db/schemas/schemas.py (hunk 2 above)
Comment (tferns) [reply db/schemas/schemas.py:58]: I'll take a deeper look at this tomorrow

Comment (tferns) [reply db/schemas/schemas.py]: Have tried to utilize fields now and removed the ugly `post_dump` stuff.  If you know a better way to do this please let me know, admittedly not too familiar with this library.

File: api/routes/assessment_routes.py (hunk 5 above)
Comment (tferns) [reply api/routes/assessment_routes.py]: Thanks for noticing that, have refactored in https://github.com/communitiesuk/funding-service-design-assessment-store/pull/50/commits/cf75286e37b0cfe1f95423e8aea0ae793ec6502d

File: db/schemas/schemas.py (hunk 3 above)
Comment (tferns) [reply db/schemas/schemas.py]: Have changed to use `only` as an argument in the constructor, again if you know a better way please let me know, thanks for the feedback 😄

File: db/schemas/__init__.py (hunk 4 above)
Comment (Adam-W1) [reply db/schemas/__init__.py:4]: just out of interest folks, is good practice pythonista stuff? My understanding was that the default 'import *' behaviour is to import all symbols that do not begin with an underscore anyway?

File: api/routes/assessment_routes.py [hunk 6]
```diff
@@ -22,3 +25,26 @@ def all_assessments_for_fund_round_id(
...
+def get_assessor_task_list_state(application_id: str) -> dict:
+    """get_assessor_task_list_state Function used by the endpoint
+    `/assessor_task_list/{application_id}`.
+
+    :param application_id: The stringified application UUID.
+    :return: A dictionary.
+    """
+
+    metadata = find_assessor_task_list_state(application_id)
+    sections, criterias = transform_to_assessor_task_list_metadata(
```
Comment (Adam-W1) [inline api/routes/assessment_routes.py:39]: 👌

File: config/mappings/assessment_mapping_fund_round.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+                    "theme_count": len(sc["themes"]),
+                    "status": "Not started",  # TODO: Get dynamic state.
+                }
+                for sc in c["sub_criteria"]
+            ],
+        }
+        for c in mapping["scored_criteria"]
+    ]
+
+    return sections, criterias
```
Comment (Adam-W1) [inline config/mappings/assessment_mapping_fund_round.py:47]: Would be good to see some testing for this

File: db/queries/assessment_records/queries.py [hunk 8]
```diff
@@ -105,3 +108,44 @@ def find_answer_by_key_runner(field_key: str, app_id: str) -> List[tuple]:
         .filter(AssessmentRecord.application_id == app_id)
         .one()
     )
+
+
+def find_assessor_task_list_state(application_id: str) -> dict:
```
Comment (Adam-W1) [inline db/queries/assessment_records/queries.py:113]: Might we want to return tasklist specific state here. For example, we only want criteria scoring and statuses, so not sure if we need the json blob of answers.

Comment (tferns) [reply db/queries/assessment_records/queries.py:113]: I think so too yeah.  Will need to find out when that work can actually be done, as I think there are in-flight stories at the moment making changes to the data model?  I'll actually ask Gio to get confirmation on this, because if not I might be able to do TDD on the current data model

File: config/mappings/assessment_mapping_fund_round.py (hunk 7 above)
Comment (tferns) [reply config/mappings/assessment_mapping_fund_round.py:47]: I've added https://github.com/communitiesuk/funding-service-design-assessment-store/pull/50/commits/f03237ce25be02fc3eadc29f7c9699c84b9310fa which sort of acts as a test/documentation for the eventually transformed structure (this is what is returned from the endpoint, the dataclass `AssessorTaskList` is built form, and used to dynamically populate the UI with config elements, and state (once part of the transformation)

File: db/queries/assessment_records/queries.py [hunk 9]
```diff
@@ -105,3 +108,44 @@ def find_answer_by_key_runner(field_key: str, app_id: str) -> List[tuple]:
...
+    assessment_record = db.session.scalar(stmt)
+
+    assessment_record_json = AssessorTaskListMetadata(
+        only=(
+            "short_id",
+            "project_name",
+            "date_submitted",
+            "workflow_status",
+            "fund_id",
+            "round_id",
```
Comment (SentientHousePlant) [inline db/queries/assessment_records/queries.py:147]: should the blob be excluded here? If so, then we shouldn't load it on line 127.

File: db/queries/assessment_records/queries.py (hunk 8 above)
Comment (tferns) [reply db/queries/assessment_records/queries.py:113]: Have created https://digital.dclg.gov.uk/jira/browse/FS-2000 for when the `score` data model changes are in place and that work can be picked up 👍

Comment (tferns) [reply db/queries/assessment_records/queries.py:113]: Have created https://digital.dclg.gov.uk/jira/browse/FS-2000 for when the score data model changes are in place and that work can be picked up 👍

**Review by SentientHousePlant — APPROVED** [review]

File: db/schemas/__init__.py (hunk 4 above)
Comment (SentientHousePlant) [reply db/schemas/__init__.py:4]: __all__ specifies what happens when you use *. While we don't use * often, many libraries do (like pdoc) and it keeps the public facing parts very clear. It's also used by pdoc to figure out what are the python things worth documenting.

**Review by SentientHousePlant — CHANGES_REQUESTED** [review]
😢 sorry to rollback my approval - but if we are adding a new route then we should test it.

- [x] Add test for new endpoint.

**Comment by tferns** [comment]
> 😢 sorry to rollback my approval - but if we are adding a new route then we should test it.

No worries makes sense!  I'll add some test coverage for this so we're more confident it works as intended.

**Comment by tferns** [comment]
@SentientHousePlant 
> 😢 sorry to rollback my approval - but if we are adding a new route then we should test it.

I've pushed https://github.com/communitiesuk/funding-service-design-assessment-store/pull/50/commits/dfc0c5e56f6b7f237041367947a9bcf97f3e0c47 with a test for the route and the database query, let me know if you have any other feedback  👍

File: config/mappings/assessment_mapping_fund_round.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    cof_r2w2_assessment_mapping as cof_r2w2,
+)
+from fsd_utils import CommonConfig
+
+fund_round_to_assessment_mapping = {
+    f"{CommonConfig.COF_FUND_ID}:{CommonConfig.COF_ROUND_2_ID}": cof_r2w2,
+}
+
+
+def transform_to_assessor_task_list_metadata(
```
Comment (Adam-W1) [inline config/mappings/assessment_mapping_fund_round.py:11]: Eventually this function will be making calls to the databse as well as config, so I wonder if this might be better placed in the route controller for returning assessment metadata

**Review by Adam-W1 — COMMENTED** [review]
look great! just a couple small comments to respond to 👍

**Comment by tferns** [comment]
> looks great! just a couple small comments to respond to 👍

Hey, thanks again for the review.  I can only see one new comment, unless you're referring to any old comments? 

I've added https://github.com/communitiesuk/funding-service-design-assessment-store/pull/50/commits/ddf7dfeab8c4f7e67726db1d56d661c76fa8f309 to better place the transformation logic.  Let me know if you have any other suggestions! Thanks

**Review by SentientHousePlant — APPROVED** [review]
Beautiful stuff Thomas! Thanks for taking the time to respond to my comments.
